Add a release page

StevenDufresne opened the pull request:
Related https://meta.trac.wordpress.org/ticket/7783.

This PR adds the "Releases" tab behind a feature flag `show_release_beta` so we can work together on the site. 


**TODO:**
Everything seems to be working properly, however, there seems to be a race condition where the first view doesn't reliably display. You can reproduce it sometimes by loading the release page often. The content will not be displayed.

StevenDufresne committed:
Squash merge try/release-page

StevenDufresne committed:
Remove releases reference.

StevenDufresne committed:
Turn around if not published.

StevenDufresne committed:
remove changelog for now.

StevenDufresne committed:
Clean up build artifacts.

StevenDufresne committed:
Update with current state.

StevenDufresne committed:
Cleanup, finish subnav.

StevenDufresne committed:
Hook up the release menu options.

StevenDufresne committed:
Make CPT queryable.

StevenDufresne committed:
Add playground url.

tellyworth committed:
Fix some bugs; add safety measures around deletion and updates

tellyworth committed:
Don't call this yet

tellyworth committed:
Handle duplicate CPTs; includes non-functional code for debugging

tellyworth committed:
Fix bug where post_parent was zeroed on update; remove debug and safety code

tellyworth committed:
Add a CLI script for creating/updating release CPTs during testing.

tellyworth commented:
@StevenDufresne I think the CPT code is kinda-sorta functional, thought not really hooked up to anything yet.

I've added a bin script you can use to (re)generate CPT entries for a specific plugin for testing purposes:

```
php bin/update-release-cpt.php --plugin=debug-bar
```

Next step is to properly hook this up to the main Plugin Directory class, and add some internal and REST APIs.

StevenDufresne committed:
Update meta details about blocks.

tellyworth committed:
Add code to create a draft post for trunk releases

Not sure this logic is quite right, since I'm only creating the draft when the version number doesn't already exist as a release.

I think this probably needs to be a revision # check, but the revision isn't yet stored for all releases; I'll look at that tomorrow.

StevenDufresne committed:
Update the release-menu-options to use navigation.

StevenDufresne committed:
Add build changes.

StevenDufresne committed:
Remove the experimental modules.

StevenDufresne committed:
Update blocks to add 'View changes' feature.

StevenDufresne committed:
Make the query work.

tellyworth committed:
Create a draft release when trunk changes exist after the most recent release

Needs more testing. Not sure if having the 'date' part is a good idea.

tellyworth committed:
Use `trunk` as the post title for draft posts

The tag and version values are both stored in postmeta if needed

StevenDufresne committed:
Get the release draft connected.

tellyworth committed:
Add a Plugin_Check class, and run it during svn import for trunk changes.

It might be ok to run it synchronously like this; otherwise I'll move it to an async job.

The Plugin_Check class isn't quite complete yet. It's copied from the code in the upload handler shortcode; we should replace that with a call to the class when it's complete.

StevenDufresne committed:
Update confirmation control.

tellyworth committed:
Fix bug in get_releases()

Turns out WP_Query uses `title` rather than `post_title` as a query param.

tellyworth committed:
Merge branch 'try/release-page' of https://github.com/StevenDufresne/wordpress.org into pr/407

tellyworth committed:
Add code to publish a release from a draft, and create the svn tag

tellyworth committed:
Add a `--publish` option to the update-release-cpt script

StevenDufresne committed:
Start working on release check blocks

tellyworth committed:
Update the release tag after successfully tagging a release

tellyworth committed:
Merge branch 'try/release-page' of https://github.com/StevenDufresne/wordpress.org into pr/407

tellyworth committed:
Update draft post and run plugin-check on any `trunk` commit

StevenDufresne committed:
Add in plugin check errors to checks section.

StevenDufresne committed:
Update the release tag and not the version for trunk urls.

StevenDufresne committed:
Fix formatting.

tellyworth committed:
Store `release_commit_log` postmeta for drafts

Will include the full commit log for trunk since the last published release.

tellyworth committed:
Merge branch 'try/release-page' of https://github.com/StevenDufresne/wordpress.org into pr/407

StevenDufresne committed:
Add a commit section.

tellyworth committed:
Store tested/requires headers during import

tellyworth committed:
Remove unused var

tellyworth committed:
Backfill release CPTs during import

This also adds a `maybe_backfill_releases()` function that could be called elsewhere eg from a one-time script.

tellyworth commented:
With the last commit [f5cc81a](https://github.com/WordPress/wordpress.org/pull/407/commits/f5cc81acb3d27296637e9ef6b70eafc4e7c40520) I was thinking about how to backfill in production, and how to eventually deal with migrating and replacing the `releases` postmeta.

Assuming we want to replace the postmeta with CPTs, it would be fairly straightforward to replace functions like `Plugin_Directory::get_releases()` with wrappers for `Plugin_Release::get_releases()` etc. And adapt  `Plugin_Directory::prefill_releases_meta()` to backfill CPTs from svn data. Then update anything that depends on the postmeta (like Release Confirmations) to work with a post object rather than the simple postmeta array format.

The main difficulty with that is migrating in such a way that everything continues to work smoothly - ie plugins that haven't been backfilled smoothly can still have releases confirmed.

I think the best way to manage that would be:

1. Have the import code backfill CPTs on the first trunk commit for each plugin. That way we'll gradually backfill actively developed plugins first. (Already done in f5cc81a)
2. When we're happy that the backfill data is good, run a one-time script to backfill other plugins. Could be done from a sandbox.
3. Then, replace the code to swap in CPTs for postmeta as outlined above.

I guess I could write and test a gated version of (3) so that it's ready to go and we just have to flip a switch.

StevenDufresne committed:
Big ui changes.

tellyworth committed:
Add untested Plugin_Publish API endpoint

As yet untested.

Note that I called it Plugin_Publish rather than Plugin_Release, so as not to confuse it with the other class which is also called Plugin_Release (we're using "Release" as both a noun and verb in different places).

tellyworth committed:
Fix dumb errors in the REST publish endpoint

tellyworth committed:
Add a quick script for displaying release info

StevenDufresne committed:
Have the save setup.

StevenDufresne committed:
Revert "Have the save setup."

This reverts commit 6ff94b3897bba96661ffce8f640ee44218592c06.

StevenDufresne committed:
Setup the saving sate.

StevenDufresne committed:
Clean up the blocks.

StevenDufresne committed:
Fix download link.

StevenDufresne committed:
Update trunk title.

StevenDufresne committed:
Add translator message.

StevenDufresne committed:
Clean up publish render.

StevenDufresne commented on wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-release.php:
```diff
@@ -0,0 +1,420 @@
+		}
+
+		// Store the commit log in postmeta. We'll only do this for drafts.
+		$last_release_revision = max( $last_release->release_revision );
```
@tellyworth I think this block needs to be inside of the `if ( $last_release )`. I just ran this with a new plugin and this errored out.

StevenDufresne replied:
My assumption is incorrect. First time drafts won't ever receive commits.

StevenDufresne replied:
Okay, so it looks like we could pass `SVN::log( $trunk_url, [ 1, 'HEAD' ] )` to get all the commits for new plugins without releases. 

However, what happens when we roll this out for old plugins that used trunk. We would pull in the whole plugin commit log. Do we care if we show something like that denotes 100 commits:

<img width="992" alt="Screenshot 2024-12-11 at 3 55 31 PM" src="https://github.com/user-attachments/assets/9ea28d5d-0269-4612-88d0-129db5680231">

To those users those commits are not new, however, it could be argued that they are new to the tag/release.

tellyworth replied:
Yeah [1, 'HEAD'] will probably work for the initial version or where no release exists.

How about if I limit the commit log meta to the most recent 5 or 10 items for each release?

StevenDufresne replied:
I think that would be less expected and maybe more confusing...? I think we maybe just list them all and see what users say?

tellyworth committed:
Store revision_final and revision_prior

Intended for use in displaying diffs

tellyworth committed:
Include commit log in all release cpts

tellyworth committed:
In get-release-info.php, display revision range and commit log

tellyworth committed:
Improve update-release-cpt.php to use svn tag data directly

This includes revision numbers, so data is more complete after running the update script now.

Also, I changed it to use the backfill function, so its behaviour should be identical to what we'll eventually use as a one-time backfill.

tellyworth committed:
Improve the get-release-info script

It now shows the commit log data if available (which it should be, if you re-run update-release-cpt.php first)

StevenDufresne committed:
Run the linter.

StevenDufresne committed:
Move the error messge to context.

StevenDufresne committed:
merger import warnings into plugin check errors.

StevenDufresne committed:
Remove accident file.

StevenDufresne committed:
More cleanup

StevenDufresne committed:
Remove the extra spaces.

StevenDufresne committed:
Change from a form to solve the redirect issue.

tellyworth committed:
Fix bug in script output

StevenDufresne committed:
Release release flags.

tellyworth committed:
Fix svn log handling

Includes the initial trunk commit log for the first release.
Also gets the order correct, and should handle plugins with no releases yet.

Thanks Steve for the suggestions!

tellyworth committed:
Merge branch 'try/release-page' of https://github.com/StevenDufresne/wordpress.org into pr/407

StevenDufresne committed:
Move away from a form and fix validation.

StevenDufresne committed:
Fix heading issues.

StevenDufresne committed:
Make the secondary font size smaller.

tellyworth committed:
Fix fatal when no release exists

tellyworth committed:
Merge branch 'try/release-page' of https://github.com/StevenDufresne/wordpress.org into pr/407

StevenDufresne committed:
Revert block-config.php

StevenDufresne committed:
Rename endpoint to publish.

StevenDufresne committed:
Refactor to save commits if there are no releases yet.

StevenDufresne committed:
Add tested for new drafts.

StevenDufresne committed:
Handle errors better.

StevenDufresne committed:
Clean up errors.

StevenDufresne committed:
Improve error messages, return errors to view.

StevenDufresne committed:
Fix the css.

StevenDufresne committed:
Fix the changelog links.

StevenDufresne committed:
Remove faileds on import warnings and plugin_check.

StevenDufresne committed:
Update heading.

StevenDufresne committed:
Darken secondary text.
